### DISCUSSION TITLE
First Day trying Avalonia, First Impressions

### DISCUSSION THREAD

**Discussion by UniqeId6542** [AvaloniaUI/Avalonia#15121] [General]
FYI - General thoughts after about a day playing with it.
 - I like the easy Dark/Light theme change in app
 - I like how the color palette can be changed
 - I like how controls follow the color palette (and other theme things - like TextControlThemeMinHeight)
 - I like having a "Compact" option for the theme
 - I like how the default project is setup
 - I like the similarity to WPF (as someone with WPF experience)

 - For a NumericUpDown, reducing the MinHeight and Height causes truncation of the control instead of resize.  I had to set TextControlThemeMinHeight to get it to resize.
 - In SimpleTheme, for a NumericUpDown, reducing the Height causes what appears to be a scroll bar for the TextBox inside of it to appear.
 - There is no web page giving all the default colors in the Fluent color palette.  (So, when I make a custom control, I won't know which are supposed to go together easily.)
 - I see a button has ButtonBackground and that ties to SystemControlBackgroundBaseLowBrush.  I also see that Fluent has BaseLow.  But I don't see how these get tied together.
 - I see that in the Dark theme, some colors in the palette change but some others stay the same.  So, I can't keep using the same colors for the other theme, and I can't use the colors differently - since the contrast between any two is not the same.
 - I want to edit the view graphically, not only by editing xaml text.
 - I want the designer to integrate with Visual Studio's Toolbox - so I can see and pick controls
 - I want the designer to integrate with Visual Studio's Properties window - so I can see a list of properties (and edit) easier.
 - I need a ListView.  Maybe I can make a DataGrid work.  (I need columns and row selection - not cell selection)  But I also will be customizing how the ListView acts, and starting with a heavily customized DataGrid seems like more work than it's worth.

From a 1-day look, Avalonia is more than half baked, though it has lots of room for editing usability improvements.  Overall, I felt OK working around the various issues.  That is, until I found no ListView.  I may yet try to modify a DataGrid, but this missing control may be what keeps me from transition to Avalonia.

**Comment by wsficke** [comment]
Please see ItemsRepeater, should solve your problems with missing ListView

https://stackoverflow.com/questions/64304901/how-to-create-a-table-like-listview-in-avaloniaui

It would be earth-shattering if Avalonia had a generative AI that could mutate your view, code-behind, and viewmodel based on GUI interactions and your patterns.  Like a more flexible version of the very rigid Windows Forms code-behind setup married to Github Copilot.  This would instantly catapult Avalonia to S-tier in the MVVM framework ranking.

**Reply by maxkatz6** [reply]
You can use ItemsRepeater for that, but you probably should use ItemsControl instead (which follows closely to what WPF did).
ItemsRepeater is getting slowly but surely replaced by ItemsControl in Avalonia.

**Comment by maxkatz6** [comment]
> I need a ListView. Maybe I can make a DataGrid work. (I need columns and row selection - not cell selection)

Yeah, you need DataGrid, as it's a row selection control, not cell selection in Avalonia.
If you would need cell selection, there is a bit more flexible TreeDataGrid.

**Reply by dbriard** [reply]
TreeDataGrid is really nice control, is it officially supported by Avalonia?
Personnally I would be very interested by a TreeListView control (simplier than TreeDataGrid). I saw a such control in UWP source code.
Use case would be to have the expand and drag'n drop capability of the TreeDataGrid, but without the columns, and with the possibility to use it horizontally, instead of vertically.

**Reply by thevortexcloud** [reply]
> TreeDataGrid is really nice control, is it officially supported by Avalonia?

Should be. It's in the official Avalonia namespace and in the Avalonia repos.

https://github.com/AvaloniaUI/Avalonia.Controls.TreeDataGrid

> Personnally I would be very interested by a TreeListView control

It sounds like you are kind of describing the tree view?

https://docs.avaloniaui.net/docs/reference/controls/detailed-reference/treeview-1

**Reply by BAndysc** [reply]
For a ListView replacement you can also check https://github.com/wieslawsoltes/DataBox which is a nice control built on top of the ListBox

**Reply by wsficke** [reply]
Wiesław's DataBox is an inspiring prototype, but it has several issues that should be resolved before productization:

1. Poor implementation of thread pool delegation.  
a. Type tunneling (passing as "object" and downcasting)
b. Missing error handling on the thread pool delegate
c. Code fix: use "await Task.Run(async ()=> ... ).ConfigureAwait(true)" or use a nested Dispatcher.UIThread callback to marshal exceptions and results with strong typing.
d. Some methods that should be "async" are not implemented as such - needs usual refactoring to support a fully async data source.
2. The async virtualizing demo is bugged, does not render rows
3. Type-specific column definitions (for DateTime, etc) are bugged - the demos declare inline templates for all the columns instead

**Reply by maxkatz6** [reply]
> TreeDataGrid is really nice control, is it officially supported by Avalonia?

Yeah, it's developed by the same people. And we actually can support it, unlikely old DataGrid, which is very hard to maintain.

**Comment by UniqeId6542** [comment]
2nd day (or so), 2nd impressions...

I tried using DataGrid.  Thanks for the suggestion.

The documentation for the syntax of a style selector is fairly good.
The documentation for tunneling events is fairly good.  (Though if I didn't know what tunneling meant or wasn't already familiar with WPF preview events, it would have been confusing.)

For my project, I chose compile bindings.  But then I found out that the Style Selector is magic string lacking compile time checks.  A misspelling in a named template items just quietly fails at runtime.

The documentation disagrees for whether more than one pseudo class can exist in a style selector.
1: https://docs.avaloniaui.net/docs/reference/styles/style-selector-syntax
There can only be one pseudo class in the selector.
2+: https://docs.avaloniaui.net/docs/reference/styles/pseudo-classes
A Control can have multiple pseudoclasses active and you can target multiple pseudoclasses with a Selector.

I tried adding a custom class MyDataGrid and referencing it in the axaml.  Through experimentation, I found that I had to first create the class and build the project; THEN I could reference it in axaml.  The error doesn't explain this; and to me, this is a compile bug.

I started with the provided example of a DataGrid with binding to FirstName.  When I changed to MyDataGrid, the binding to FirstName creates a compile time error saying that the property doesn't exist in MainViewModel.  (It should be looking at Person.)

I want the DataGrid to act like a ListView - so right-key should scroll right instead of select a cell to the right.  I was able to use FindDescendantOfType on the DataGrid to get a ScrollViewer.  But calling LineRight (or PageRight) on that does nothing.

I want the DataGrid to act like a ListView - so I found how to hide the focus visual on the selected cell.  The suggested style selector was:
"DataGrid:focus DataGridCell:current /template/ Grid#FocusVisual"
Through experimentation I found that these also work:
"DataGrid:focus DataGridCell /template/ Grid#FocusVisual"
"DataGridCell:current /template/ Grid#FocusVisual"
"DataGrid DataGridCell:current /template/ Grid#FocusVisual"
To me, this implies that ":focus" is optional and ":current" is also optional.  But, this does NOT work:
"DataGrid DataGridCell /template/ Grid#FocusVisual"

How am I supposed to know that the suggested style selector would work without asking someone?  From the documentation, I can figure out: "When the DataGrid has the pseudoClass 'focus' true consider its child class (of any depth) DataGridCell, when that has the pseudoClass 'current' true; then go to the template and find the Grid named FocusVisual".
I found 'current' as a pdeudoClass for DataGridCell.  But the documentation doesn't say what it means.
Am I looking at the template of the DataGrid or DataGridCell?
I checked the DataGrid and DataGridCell classes, but neither define a TemplatePart of FocusVisual.  How am I supposed to know that it exists?  If it's not documented, does that mean that it could be removed at any random time?

Avalonia is feeling more half baked.
I may yet try TreeDataGrid.  Or, maybe I'll figure out ways around the current DataGrid issues.

**Reply by thevortexcloud** [reply]
> For my project, I chose compile bindings. But then I found out that the Style Selector is magic string lacking compile time checks. A misspelling in a named template items just quietly fails at runtime.

There is actually some stuff in 11.1 to help with this that should raise compile warnings/errors if you define a template without the required template parts. It probably wont help with a selector specifically, but it's start. The problem with template parts is a lot of them are optional, so there is no guarantee they will even exist in a control. That said the style selector is validated at compile time generally if you are targeting controls directly, as it's much easier to check if a property exists or not in that scenario.

> I tried adding a custom class MyDataGrid and referencing it in the axaml. Through experimentation, I found that I had to first create the class and build the project; THEN I could reference it in axaml. The error doesn't explain this; and to me, this is a compile bug.

You should not need to do this. If you have an error in your code somewhere it can stop the source generators from working which can then stop everything else working though. But it's not always obvious.

> How am I supposed to know that the suggested style selector would work without asking someone? 

Officially? The same way as you do in WPF. By reading the source code. All of the pseudo selectors and template part names are defined in attributes in the control class. EG 

https://github.com/AvaloniaUI/Avalonia/blob/76a1c70758884abcebc9e3665aa08290a841efa4/src/Avalonia.Controls.DataGrid/DataGridRow.cs#L26-L32

That said I have also said in the past that forcing people to check source/decompile Avalonia is not a great option. I was never happy that WPF made people do that too. There was some discussions about this awhile back to auto generate docs for Avalonia based on these attribute values but I am not sure if that went anywhere.

**Reply by StefanKoell** [reply]
Regarding style selectors I find the DevTool helpful.

**Reply by UniqeId6542** [reply]
Thanks for your reply.

> You should not need to do this. If you have an error in your code somewhere it can stop the source generators from working which can then stop everything else working though. But it's not always obvious.

That could be.  I can't currently reproduce it, so maybe a namespace was off.

> Officially? The same way as you do in WPF. By reading the source code. All of the pseudo selectors and template part names are defined in attributes in the control class. EG

That's the issue.  FocusVisual is NOT a template part.  It is defined for the DataGridCell ONLY in xaml.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls.DataGrid/Themes/Fluent.xaml
It can't be found by going to the definition of a class.
It's not declared publicly, so it could in theory change without notice.
I could look into overriding DataGridCellFocusVisualPrimaryBrush, but that also appears to be non-public.

Do you have any suggestion for the FirstName binding issue when I switch to a custom class derived from DataGrid?

**Reply by stevemonaco** [reply]
> There can only be one pseudo class in the selector.

That doc is wrong on this point. You can have multiple pseudoclasses, even on the same element. I'll PR a change. Please report future doc issues to https://github.com/AvaloniaUI/avalonia-docs

> To me, this implies that ":focus" is optional and ":current" is also optional. But, this does NOT work:
"DataGrid DataGridCell /template/ Grid#FocusVisual"
How am I supposed to know that the suggested style selector would work without asking someone?

Checking DevTools to see what styles are applied to `Grid#FocusVisual` at runtime and which are inactive. If you've gotten this far, you should have already read the `ControlTemplate` for `DataGridCell`. You can see that the non-working selector doesn't have a pseudoclass -- this is important. Pseudoclasses are conditional activators that increase the priority. You can read https://docs.avaloniaui.net/docs/guides/styles-and-resources/setter-precedence for details.

> That's the issue. FocusVisual is NOT a template part. It is defined for the DataGridCell ONLY in xaml.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls.DataGrid/Themes/Fluent.xaml
It can't be found by going to the definition of a class.
It's not declared publicly, so it could in theory change without notice.
I could look into overriding DataGridCellFocusVisualPrimaryBrush, but that also appears to be non-public.

This is common in XAML. Controls are lookless, so they require the minimum number of parts to satisfy the core behavioral requirements. Any style creator can create as many extra components as they want. This ultimately means that full understanding requires reading the control's public API **and** the `ControlTheme` for the specific theme you're using. `DataGrid` is probably the most complex control in the repo.

**Reply by thevortexcloud** [reply]
> Do you have any suggestion for the FirstName binding issue when I switch to a custom class derived from DataGrid?

You can either use a reflection binding or you can explicitly set the data type of the item when setting up the columns. DataGrid is in general kind of annoying to setup with compiled bindings and there is often a lot of repetition.

> I could look into overriding DataGridCellFocusVisualPrimaryBrush, but that also appears to be non-public.

Dynamic resources can actually be overridden. That's one of the main points of them. 

https://docs.avaloniaui.net/docs/guides/styles-and-resources/resources#resource-priority

WPF's docs also go into more detail about how and where you would use them.

https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/dynamicresource-markup-extension?view=netframeworkdesktop-4.8

Since the DataGrid theme is typically defined at the application level, anything defined closer to the DataGrid will have priority.

**Comment by robloo** [comment]
Just some quick comments based on your assessment:

> Avalonia is feeling more half baked.

Avalonia 0.10.x was pretty solid with few issues. This is the version that was first investigated on my end and I was amazed at what it could do already. The problem was it was still missing a lot of key functionality other XAML frameworks had -- and that existing apps need.

Avalonia 11.0 basically fixed all the major missing pieces I am aware of on my end. This is where I jumped in and helped where I could to close the feature gap with WinUI as well. It's not widely stated but it actually pretty smooth now to go from WinUI -> Avalonia+FluentAvalonia where before only WPF->Avalonia was widely considered. However, 11.0 was less solid than 0.10.x out of the gate simply due to the large number of new features/changes. Everyone has been working on those since. Subsequent 11.x releases are getting more and more stable approaching what was in 0.10.x but now with all the new features.

As someone who was working with UWP/WinUI and then switched over to Avalonia my impression is Avalonia already is better than that tech stack -- by quite a bit in some areas. You can do a lot more in Avalonia (and WPF) that was lost in UWP. It's also much more stable and issues actually get fixed (slowly is better than never which was the case with WinUI for years).

Personally, I expect Avalonia to be at the level of WPF in terms of stability within the 11.x series. By 12.0 Avalonia will start to pull ahead and really break new ground.

All of this says nothing about the fact you can do a lot in Avalonia that just wasn't possible or was extremely difficult in other XAML frameworks.
